### PR TITLE
Adding recipe for the Insight Toolkit Development

### PR DESCRIPTION
**Pull request by blowekamp**
This package contains ITK static libraries along with the headers. It
is suitable for other packages to use with out a run time dependency
on this library but the trasitive dependencies on the system libraries
such as libpng and libtiff may be required.

### CODE REVIEW

File: recipes/itk-dev/.#bld.bat [hunk 1]
```diff
@@ -0,0 +1 @@
+blowekamp@lhcp-rh7.nlm.nih.gov.79853:1490965283
```
Comment (jakirkham) [inline recipes/itk-dev/.#bld.bat]: Do we need this file?

File: recipes/itk-dev/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{% set version = "4.11.0" %}
+
+package:
+  name: itk-dev
+  version: "{{version}}"
+
+source:
+  git_url: git://itk.org/ITK.git
+  git_tag:  v{{version}}
```
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: We generally prefer using a tarball to a git clone. Is there a tarball or archive that we could use?

File: recipes/itk-dev/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    - vc 10  # [win and py34]
+    - vc 14  # [win and py35]
+
+test:
+  commands:
+    - echo "Testing..."
+
+about:
+  home: http::/www.itk.org
+  license: Apache 2.0
```
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: Is there a license file included?

File: recipes/itk-dev/.#bld.bat (hunk 1 above)
Comment (blowekamp) [inline recipes/itk-dev/.#bld.bat]: Extraneous temporary file.

File: recipes/itk-dev/meta.yaml (hunk 2 above)
Comment (blowekamp) [inline recipes/itk-dev/meta.yaml]: We should be able to use the one auto generated from reliable Github, as opposed the manually created one on SourceForge.

File: recipes/itk-dev/meta.yaml (hunk 3 above)
Comment (blowekamp) [inline recipes/itk-dev/meta.yaml]: The following files are installed/packaged:
```
$ ls -R share/doc/ITK-4.11/
share/doc/ITK-4.11/:
itksys  LICENSE  NOTICE  README.txt

share/doc/ITK-4.11/itksys:
Copyright.txt
```

File: recipes/itk-dev/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{% set version = "4.11.0" %}
+
+package:
+  name: itk-dev
+  version: "{{version}}"
+
+source:
+  url: https://github.com/InsightSoftwareConsortium/ITK/archive/v{{version}}.tar.gz
```
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: As the linter says, we need a checksum. Preferably `sha256`. Preferably not `md5`. Though others could work too.

File: recipes/itk-dev/meta.yaml (hunk 3 above)
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: So I guess pick one and add it with [`license_file`]( https://conda.io/docs/building/meta-yaml.html#license-file ). Would guess `LICENSE` would be the one to go with, but leaving it up to you.

File: recipes/itk-dev/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+
+build:
+    number: 0
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: Would add `toolchain` here.

File: recipes/itk-dev/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,46 @@
+{% set version = "4.11.0" %}
+
+package:
+  name: itk-dev
```
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: Should we call it `itk-dev` or just `itk`?

File: recipes/itk-dev/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    - libpng   >=1.6.27,<1.7
+    - jpeg     9*
+    - zlib     1.2.*
+    - vc 9  # [win and py27]
+    - vc 10  # [win and py34]
+    - vc 14  # [win and py35]
+
+test:
+  commands:
+    - echo "Testing..."
```
Comment (jakirkham) [inline recipes/itk-dev/meta.yaml]: Could we add some file tests here? Checking for headers, libraries, running CLIs with `--help`, etc.

File: recipes/itk-dev/meta.yaml (hunk 6 above)
Comment (blowekamp) [inline recipes/itk-dev/meta.yaml]: As this package contains static libraries, and headers it is not needed for run-time usage. It seemed more like a linux `devel` package. If there is not existing convention for this type of package, just `itk` is fine if that is the practice for condo-forge packages.

File: recipes/itk-dev/meta.yaml (hunk 7 above)
Comment (blowekamp) [inline recipes/itk-dev/meta.yaml]: I was thinking of configuring minimal CMake project which uses ITK. But we'll come up with a couple things, not that we have enough time in the build.

**Comment by isuruf** [comment]
What's the reason for having static libraries? You can just as well have shared libraries and let the package manager handle the dependencies for you.

Also, there are many bundled dependencies that is being built, but are `conda-forge` packages. Eg: OpenJPEG, GDCM, HDF5.

**Comment by blowekamp** [comment]
This is derived from the SimpleITK package that we post on anaconda cloud: https://github.com/SimpleITK/SimpleITKCondaRecipe Here we created a monolithic package that is easily distributed as a binary package.

Regarding dependencies, my early experience with build a conda package with many dependencies was that it was frequently became incompatible with other packages. Additionally, figuring to the build version and the compatible runtime version range was too difficult. I suppose with condo-forge automatically generating and maintaining package depend this should not be an issue.
- I will revisit the list of `system` libraries ITK support with condo-forge. Early testing on linux indicate that a system HDF5 work but GDCM/OpenJPEG did not

Regarding static libraries, this is what was done with the SimpleITK package. There are a couple reasons due to the fact that ITK is mostly a template header library. This is done to help manage the 10's of millions of systems that can be created with templates. It creates smaller libraries, faster load times, along with reduces interoperability problems with symbols visibility. Additionally, since ITK is a template library a change it its version requires recompilation of the dependers.

**Comment by blowekamp** [comment]
I added `hdf5` and `expect` as from condo-forge as system libraries. Enabling GDCM currently causes a compilation error:
FAILED: Modules/IO/GDCM/src/CMakeFiles/ITKIOGDCM.dir/itkGDCMImageIO.cxx.o 
/usr/lib64/ccache/g++    -I/home/blowekamp/miniconda2/conda-bld/itk-dev_1492713557094/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/include/gdcm-2.6 -IModules/ThirdParty/KWIML/src -I../Modules/ThirdParty/KWIML/src -IModules/ThirdParty/KWSys/src -I../Modules/ThirdParty/VNL/src/vxl/v3p/netlib -I../Modules/ThirdParty/VNL/src/vxl/vcl -I../Modules/ThirdParty/VNL/src/vxl/core -IModules/ThirdParty/VNL/src/vxl/v3p/netlib -IModules/ThirdParty/VNL/src/vxl/vcl -IModules/ThirdParty/VNL/src/vxl/core -I../Modules/ThirdParty/VNLInstantiation/include -IModules/Core/Common -I../Modules/Core/Common/include -IModules/IO/ImageBase -I../Modules/IO/ImageBase/include -I../Modules/IO/GDCM/include -I../Modules/ThirdParty/VNL/src/vxl/core/vnl/algo -I../Modules/ThirdParty/VNL/src/vxl/core/vnl -DBOOST_MATH_DISABLE_FLOAT128 -m64  -Wall -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -funit-at-a-time -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Woverloaded-virtual -Wstrict-null-sentinel  -Wno-array-bounds -O3 -DNDEBUG -fPIC -MD -MT Modules/IO/GDCM/src/CMakeFiles/ITKIOGDCM.dir/itkGDCMImageIO.cxx.o -MF Modules/IO/GDCM/src/CMakeFiles/ITKIOGDCM.dir/itkGDCMImageIO.cxx.o.d -o Modules/IO/GDCM/src/CMakeFiles/ITKIOGDCM.dir/itkGDCMImageIO.cxx.o -c /home/blowekamp/miniconda2/conda-bld/itk-dev_1492713557094/work/ITK-4.11.0/Modules/IO/GDCM/src/itkGDCMImageIO.cxx
/home/blowekamp/miniconda2/conda-bld/itk-dev_1492713557094/work/ITK-4.11.0/Modules/IO/GDCM/src/itkGDCMImageIO.cxx:41:29: fatal error: gdcmImageHelper.h: No such file or directory
 #include "gdcmImageHelper.h"

**Comment by isuruf** [comment]
I think that's because the headers are not installed here, https://github.com/conda-forge/gdcm-feedstock/blob/master/recipe/build.sh#L26
Can you raise an issue or send a PR to fix it?

**Comment by bluescarni** [comment]
I have a recipe I have been testing locally in the last few days for ITK as well:

https://github.com/bluescarni/staged-recipes/tree/itk

This one takes essentially the default ITK build, disable the tests which take ages to compile, and enables all the system dependencies I could identify on conda-forge. CI seems fine on all three platforms.

Should we try to join them somehow?

**Comment by bluescarni** [comment]
Some passing builds:

https://circleci.com/gh/bluescarni/staged-recipes/5
https://ci.appveyor.com/project/bluescarni/staged-recipes/build/1.0.2
https://travis-ci.org/bluescarni/staged-recipes/builds/224322336

I ran into the same issue with the GCDM system library.

**Comment by blowekamp** [comment]
I apologize for having difficulties with this pull request, in getting the proper code into git.

Disabling the test is important. I think it should all be working now.

Most of the additional CMake configuration variables are make certain default options explicit, such as:  `-D BUILD_EXAMPLES:BOOL=OFF`,   `-D ITK_USE_KWSTYLE:BOOL=OFF`, `-D ITK_BUILD_DEFAULT_MODULES:BOOL=ON`

I find the follow option is needed to ensure that the conda libraries are detected when building locally:  `-D "CMAKE_SYSTEM_PREFIX_PATH:PATH=%PREFIX%/Library"`

There are addition configuration options that are good to apply to remove old 3.0 code, and enable the commonly used "Review" module: `-D ITK_LEGACY_REMOVE:BOOL=ON`, `-D Module_ITKReview:BOOL=ON`

In your yaml you have:
```
  skip: True  # [win and py<34]
  features:
    - vc10  # [win and py34]
    - vc14  # [win and py>=35]
```
This may be required here.



Are there any specific changes to the configurations you'd suggest?

Additionally there are the following open issues:
-Shared libraries?
- Use conda GDCM - I don't think this is a good option because condo gdcm is a python package with python version dependencies, for this package we don't want to have a dependency on the python version.

**Comment by jakirkham** [comment]
In the case of name conflicts, we aim for as little disruption as possible. So given there is already a `gdcm` package for Python, we should disambiguate the library somehow. Prefixing with `lib` is a common choice. So could do `libgdcm`. Other options have included prefixing or postfixing language details (e.g. `cpp-gdcm` or `gdcm-cpp` or using `cxx`). Personally I recommend `libgdcm`. Thoughts from others?

**Comment by bluescarni** [comment]
Hello @blowekamp 

I don't have experience with using ITK myself, I needed the recipe because software I am working on relies on ITK as a dependency.

What I did was just to take the default CMake build, and implement a few alterations to make it play nicely with the rest of the conda environment:

- use conda-provided dependencies if possible (currently HDF5, jpeg, png, tiff, expat, zlib),
- build shared libraries.

I am not sure what to do about the testing situation because the full test suite will timeout the CI, and I see no option in CMake to enable a reduced version of it.

Regarding older legacy code, if it does not cost too much in terms of compilation time my suggestion would be to leave it in, to maximise the usefulness of the recipe. Shared libraries would also seem the way to go in my opinion.

Regarding the GCDM situation, I guess maybe we could use the builtin version in ITK for the time being and see what kind of fixes we can apply to the conda-forge package?

**Comment by blowekamp** [comment]
@bluescarni  Sorry! I ment to add you as a co-author for the last patch: me cupa

We should be really close to a good recipe now. ( Forgot to rename package to `itk`)

I am already 3-layers deep with trying to get my  SimpleITK python package into conda-forge. I am reluctant to dive another layer to create a libgdcm package. This is certainly something that should be done, but I hope it is not a block for itk.

**Comment by jakirkham** [comment]
Can you build everything you need with GDCM support disabled? If so, I'd recommend doing that and leaving an issue on the feedstock to fix that later.

File: recipes/itk/bld.bat [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    -D ITK_USE_SYSTEM_HDF5:BOOL=ON ^
+    -D ITK_USE_SYSTEM_JPEG:BOOL=ON ^
+    -D ITK_USE_SYSTEM_PNG:BOOL=ON ^
+    -D ITK_USE_SYSTEM_TIFF:BOOL=ON ^
+    -D ITK_USE_SYSTEM_ZLIB:BOOL=ON ^
+    -D ITK_USE_KWSTYLE:BOOL=OFF ^
+    -D ITK_BUILD_DEFAULT_MODULES:BOOL=ON ^
+    -D Module_ITKReview:BOOL=ON ^
+    -D "CMAKE_SYSTEM_PREFIX_PATH:PATH=%PREFIX%/Library" ^
+    -D "CMAKE_INSTALL_PREFIX=%PREFIX%" ^
```
Comment (isuruf) [inline recipes/itk/bld.bat]: Change this to `%PREFIX%/Library` or `%LIBRARY_PREFIX%`. (Both are the same)

File: recipes/itk/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    - example
+  commands:
+    - test -d $PREFIX/include/ITK*                        # [unix]
+    - if not exist %LIBRARY_INC%\\ITK* exit 1             # [win]
+    - itkTestDriver -- echo "Runtime ITK OK"
+    - cmake -D "CMAKE_SYSTEM_PREFIX_PATH:FILEPATH=${PREFIX}" ./example  # [unix]
+    - cmake -D "CMAKE_SYSTEM_PREFIX_PATH:FILEPATH=%PREFIX%" ./example   # [win]
+
+about:
+  home: http::/www.itk.org
```
Comment (isuruf) [inline recipes/itk/meta.yaml]: `http://`

File: recipes/itk/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    - vc 9   # [win and py27]
+    - vc 10  # [win and py34]
+    - vc 14  # [win and py>=35]
+
+test:
+  files:
+    - example
+  commands:
+    - test -d $PREFIX/include/ITK*                        # [unix]
+    - if not exist %LIBRARY_INC%\\ITK* exit 1             # [win]
```
Comment (isuruf) [inline recipes/itk/meta.yaml]: check for the libraries like
```
- test -f $PREFIX/lib/libITKCommon-{{ ".".join(version.split(".")[:2]) }}.so        # [linux]
- test -f $PREFIX/lib/libITKCommon-{{ ".".join(version.split(".")[:2]) }}.dylib    # [osx]
```

Comment (isuruf) [inline recipes/itk/meta.yaml]: You are doing an example build. This is not needed

File: recipes/itk/bld.bat (hunk 8 above)
Comment (blowekamp) [inline recipes/itk/bld.bat]: Only on Window or linux/osx too?

File: recipes/itk/meta.yaml (hunk 10 above)
Comment (blowekamp) [inline recipes/itk/meta.yaml]: Agreed this is not needed, with the exists tests.

The CMake example configuration mainly checks the proper existence of CMake files referencing library files, it does not check for headers.

Also running "itkTestDriver", does a run time check some ITK libraries too.

File: recipes/itk/bld.bat (hunk 8 above)
Comment (isuruf) [inline recipes/itk/bld.bat]: only on windows

File: recipes/itk/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    number: 0
+    skip: True   # [win and py36]
+    features:
+    - vc9   # [win and py27]
+    - vc10  # [win and py34]
+    - vc14  # [win and py>=35]
+
+requirements:
+  build:
+    - toolchain
```
Comment (jakirkham) [inline recipes/itk/meta.yaml]: Please add the following.

```yaml
    - python  # [win]
```

Comment (blowekamp) [inline recipes/itk/meta.yaml]: Why is this needed? What will it do?

**Comment by blowekamp** [comment]
The build finally seems to be work now. Thanks for all the advice, help and patience!

File: recipes/itk/meta.yaml (hunk 11 above)
Comment (jakirkham) [inline recipes/itk/meta.yaml]: Note that all of the `vc` portions rely on selectors related to the version of `python` used. As such we need this to make sure those selectors work. Also if we don't have this listed, then we get multiple rebuilds of the package on Unix for each Python version, which we don't need.

**Comment by blowekamp** [comment]
OK, the travis error is this:
```
The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over).
```
Need to add a flag to strip to make it less verbose.

**Comment by bluescarni** [comment]
@blowekamp what I did in my recipe was just to add ``> /dev/null`` to the installation step (which is the one that generates the most output).

**Comment by blowekamp** [comment]
@bluescarni  I'm just disabling stripping to get this package in.

I don't understand why the build on Travis is now timing out. There is no substantiate change in the script since it was just taking 20-30 min for the build.

**Comment by jakirkham** [comment]
Maybe slow downloads? Haven't been following that closely though. Did restart Travis CI. So that should help. If the issue persists, might be worth investigating.

File: recipes/itk/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    - python
```
Comment (blowekamp) [inline recipes/itk/meta.yaml]: Missing the # [win]!

Comment (jakirkham) [inline recipes/itk/meta.yaml]: Yep, that would do it.

**Comment by blowekamp** [comment]
There we go, all green! Any outstanding issues?

**Comment by blowekamp** [comment]
@bluescarni @thewtex @hjmjohnson @fbudin69500

Any other volunteers for maintainers?

**Comment by isuruf** [comment]
I'm going to merge this. You can add more maintainers in the feedstock if anybody volunteers.

**Comment by isuruf** [comment]
@blowekamp, can you create an issue in the feedstock about using gdcm and other dependencies?

**Comment by blowekamp** [comment]
Please see: conda-forge/gdcm-feedstock#3 I believe the other recommendation was to create a "libgdcm" package to be python version independent.

I am not aware of other dependency issues?

**Comment by isuruf** [comment]
In the log I see that it's building a BLAS package using f2c as well.

**Comment by bluescarni** [comment]
@blowekamp I can assist with maintenance if you like.

**Comment by blowekamp** [comment]
Those can not be used as "system" libraries. The code was converted with `f2c` and modified a long time ago.

You can see what other packager have done for dependencies here: https://packages.debian.org/stretch/libinsighttoolkit4.10

- In this configuration we did not enable FFTW.
- libnifiti2?
- libdoubleconvert?
- There use to be a VXL dependency, but they seem to be using ITK's now. VXL is what has a converted BLAS library.

I did not see a libdoubleconvert  or a libnifti2 in conda-forge.

**Comment by isuruf** [comment]
Ah, then gdcm is the only dependency we can conda-forge

**Comment by thewtex** [comment]
:tada: Well done!

Like `libgdcm`, maybe this should be renamed to `libitk` or `libinsighttoolkit`.
